rust/nasl-builtin-network/src/socket.rs

```diff
@@ -0,0 +1,731 @@
+            NaslSocket::Tcp(conn) => {
+                let mut old = None;
+                if let Some(timeout) = timeout {
+                    old = conn.socket.read_timeout().unwrap();
+                    conn.socket
+                        .set_read_timeout(Some(Duration::from_secs(timeout as u64)))?;
+                }
+                // TODO: process min for magic read function"
+                match conn.socket.read(data.as_mut_slice()) {
+                    Ok(size) => {
+                        data.truncate(size);
+                        ret = Ok(NaslValue::Data(data));
+                    }
+                    Err(e) => ret = Err(e.into()),
+                };
+
+                if let Some(timeout) = old {
+                    conn.socket.set_read_timeout(Some(timeout))?;
+                }
+
+                ret
+            }
+            NaslSocket::Udp(conn) => {
+                let mut old = None;
+                if let Some(timeout) = timeout {
+                    old = conn.socket.read_timeout().unwrap();
+                    conn.socket
+                        .set_read_timeout(Some(Duration::from_secs(timeout as u64)))?;
+                }
+
+                let mut result = conn.socket.recv(data.as_mut_slice());
+
+                for _ in 0..5 {
+                    match result {
+                        Ok(size) => {
+                            data.truncate(size);
+                            ret = Ok(NaslValue::Data(data));
+                            break;
+                        }
+                        Err(e) => match e.kind() {
+                            io::ErrorKind::TimedOut => {
+                                conn.socket.send(&conn.buffer)?;
+                            }
+                            kind => {
+                                ret = Err(FunctionErrorKind::IOError(kind));
+                                break;
+                            }
+                        },
+                    };
+
+                    result = conn.socket.recv(data.as_mut_slice());
+                }
+                if let Some(timeout) = old {
+                    conn.socket.set_read_timeout(Some(timeout))?;
+                }
+                ret
+            }
+            NaslSocket::Close => Err(FunctionErrorKind::WrongArgument(
+                "the given socket FD is already closed".to_string(),
+            )),
+        }
+    }
+
+    /// Open a KDC socket. This function takes no arguments, but it is mandatory that keys are set. The following keys are required:
+    /// - Secret/kdc_hostname
+    /// - Secret/kdc_port
+    /// - Secret/kdc_use_tcp
+    fn open_sock_kdc(
+        &self,
+        _: &Register,
+        context: &Context,
+    ) -> Result<NaslValue, FunctionErrorKind> {
+        let hostname = match get_kb_item(context, "Secret/kdc_hostname")? {
+            NaslValue::String(x) => x,
+            x => {
+                return Err(FunctionErrorKind::Diagnostic(
```

Comment:
if you rewrite get_kb_item to return a `Result<Option<NaslValue>, FunctionErrorKind>` instead of `Result<NaslValue, FunctionErrorKind>` we could differentiate if it was set to a wrong type or not set at all.

Reply:
That even had the potential to create a second method `get_required_kb_item` with a &str parameter to build the FunctionErrorKind accordingly.